### ISSUE TITLE
Crash when using String() instead of fmt.Sprintf (S1025)

### ISSUE THREAD

**Issue by Deleplace** [dominikh/go-tools#818]
```
$ staticcheck -version
staticcheck 2020.1.5
$ staticcheck -debug.version
staticcheck 2020.1.5

Compiled with Go version: go1.14.7
Main module:
	honnef.co/go/tools
Dependencies:
	github.com/BurntSushi/toml@v0.3.1 (sum: h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=)
	golang.org/x/tools@v0.0.0-20191130070609-6e064ea0cf2d (sum: h1:/iIZNFGxc/a7C3yWjGcnboV+Tkc7mxr+p6fDztwoxuM=)
$ go version
go version go1.14.7 linux/amd64
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/deleplace/.cache/go-build"
GOENV="/home/deleplace/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/deleplace/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build871482799=/tmp/go-build -gno-record-gcc-switches"
```

This code in my project:
```
u := user.Current(ctx)
email := fmt.Sprintf("%s", u)
```
generates this staticcheck warning:
```
should use String() instead of fmt.Sprintf (S1025)
```

When I apply the advice:
```
u := user.Current(ctx)
email := u.String()
```

my program crashes. This is because
- `fmt.Sprintf("%s", u)` is always able to process a nil `u`
- `(*user.User).String()` is not designed to work with a nil receiver, in which case it panicks
- the type `user.User` is in a 3rd party lib that I don't want to modify to handle a nil receiver 

While it is nice when a method properly handles a nil receiver, I don't think it's really mandatory. Some libraries may find it very sensible that calling methods on a nil pointer of their types is a programmer error, and that panicking is appropriate.

I'd like to figure out how we should ideally deal with this:
- determine that the library is "faulty" because calling `String()` should not panic, even when the receiver is nil,
- determine that my code is "faulty" because I need to check for nil by myself, and apply a different logic in the "if" and "else" bodies,
- restrict S1025 to non-reference values, where `String()` is less expected to panic (though it may still panic because of a nil pointer field inside a struct, of because of invariant violation detection in the `String()` implementation),
- restrict S1025 to cases where deep analysis of the implementation of `String()`, and of the specific receiver value at call site, show that either the receiver is not null, or `String()` handles a nil receiver without panicking.

Thoughts?

**Comment by ainar-g** [comment]
<p>The&#xA0;<a
href="https://golang.org/pkg/fmt/#hdr-Printing">documentation</a>
of&#xA0;<code>fmt</code> mentions that <code>fmt.Stringer</code>
implementations can panic, although&#xA0;it doesn't <q>bless</q>
the&#xA0;practice&#xA0;either:</p>

<blockquote>
<p>If an Error or String method triggers a panic when called by a print
routine, the fmt package reformats the error message from the panic,
decorating it with an indication that it came through the fmt
package.  […]</p>
</blockquote>

<p>So&#xA0;the&#xA0;check as&#xA0;it is&#xA0;now is
probably&#xA0;incomplete.</p>

<p><sup>(With that said, what <strong>is</strong> the email of no
user&#x202F;:-)&#x202F;?)</sup></p>

**Comment by dominikh** [comment]
> determine that the library is "faulty" because calling String() should not panic, even when the receiver is nil,

I would not expect most String methods to handle a nil receiver, and I wouldn't call them faulty.

> determine that my code is "faulty" because I need to check for nil by myself, and apply a different logic in the "if" and "else" bodies,

I would certainly argue that your current code is not very self-explanatory, relying on fmt.Sprintf to handle a nil user for you. It also seems unlikely to me that `<nil>` is a valid email address. However, that is specific to this example. I am not sure if that argument holds for all code in existence.

> restrict S1025 to non-reference values, where String() is less expected to panic 

You correctly note yourself that this wouldn't fix all instances of the issue. At the same time, it _would_ prevent the check from checking the vast majority of calls, as most values will be pointers, rendering the check mostly useless.

> restrict S1025 to cases where deep analysis of the implementation [...]

Basically suffers from the same issue of making the check useless.

I only see two viable options here:
1. live with the false positive
2. remove the check because it is wrong

I am currently leaning towards number 2. 

A third option would be turning this into a "quickfix" check, which is a (future) category of checks that only offers optional code changes, inside your editor. But even then I'm not sure we should make a suggestion that can change the meaning of code this drastically.
